xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc

```diff
@@ -152,70 +156,88 @@ bool IsAlignedSlice(const HloInstruction* slice) {
+// while loop.
+bool IsValueFunctionOfLoopIndvar(const HloInstruction& value,
+                                 CallGraph* call_graph) {
+  std::vector<HloInstruction*> callers =
+      call_graph->GetComputationCallers(value.parent());
+  if (callers.size() != 1) {
+    VLOG(5) << "Computation has multiple callers: "
+            << absl::StrJoin(callers, ",",
+                             [](std::string* out, const HloInstruction* instr) {
+                               out->append(instr->name());
+                             });
     return false;
   }
-
-  // Next, we check that the parameter used in offset computation is the loop
-  // induction variable
-  int64_t param_idx = offset.operand(2)->tuple_index();
-  const HloInstruction* root = offset.parent()->root_instruction();
-  if (root->opcode() != HloOpcode::kTuple) {
+  HloInstruction* while_op = callers[0];
+  if (while_op->opcode() != HloOpcode::kWhile) {
+    VLOG(5) << "Computation caller is not while, it is "
+            << while_op->ToString();
+    return false;
+  }
+  HloComputation* while_body = while_op->while_body();
+  std::optional<int64_t> tuple_idx = GetLoopInductionVarTupleIdx(while_op);
+  if (!tuple_idx.has_value()) {
+    VLOG(5) << "Induction variable tuple index is nullopt";
     return false;
   }
-  // Check the update operation
-  const HloInstruction* updated_var =
-      offset.parent()->root_instruction()->operand(param_idx);
-  if (!Match(updated_var, m::Add(m::GetTupleElement(m::Parameter(0), param_idx),
-                                 m::ConstantScalar(1)))) {
+  if (while_body->num_parameters() != 1) {
+    VLOG(5) << "Expected only one parameter for while body: "
+            << while_op->ToString();
     return false;
   }
-  // Check that the condition considers this.
-  const HloInstruction* condition_root =
-      while_instr->while_condition()->root_instruction();
-  if (!Match(condition_root,
-             m::Lt(m::GetTupleElement(m::Parameter(0), param_idx),
-                   m::ConstantScalar(trip_count)))) {
+  if (!while_body->parameter_instruction(0)->shape().IsTuple()) {
+    VLOG(5) << "Expected while body param to be a tuple, found "
+            << while_op->ToString();
     return false;
   }
-  // Check init
-  const HloInstruction* init_loop_iter =
-      while_instr->operand(0)->operand(param_idx);
-  if (!Match(init_loop_iter, m::ConstantScalar(0))) {
+  HloInstruction* indvar = hlo_query::GetUniqueGteInstruction(
+      while_body->parameter_instruction(0), *tuple_idx);
+  if (!indvar) {
+    VLOG(5) << "Unable to find unique GTE for while induction variable idx: "
+            << *tuple_idx << ", while op: " << while_op->ToString();
     return false;
   }
+  const HloInstruction* update =
+      while_body->root_instruction()->operand(*tuple_idx);
+  std::unique_ptr<HloModule> update_module = ExtractModule(update);
+  std::unique_ptr<HloModule> value_module = ExtractModule(&value);
```

Comment:
Instead of creating two new modules, we could just walk backwards from `update` and `value` all the way to the `while_body`'s tuple parameter? If we got there by any other path than through the right `GTE`, then we return `false`. Similarly if one of the instructions is a `replica-id` or a `partition-id`.

That'd be cheaper, and not very complex, I think. WDYT?

Reply:
Yes, thanks for the suggestion. Implemented it that way.